server/src/main/java/io/deephaven/server/arrow/ExchangeMarshaller.java

```diff
@@ -0,0 +1,122 @@
+     */
+    @Nullable
+    static ExchangeMarshaller getMarshaller(final Object export, final Collection<ExchangeMarshaller> marshallers) {
+        for (final ExchangeMarshaller marshaller : marshallers) {
+            if (marshaller.accept(export)) {
+                return marshaller;
+            }
+        }
+        return null;
+    }
+
+    /**
+     * The priority of this marshaller.
+     *
+     * <p>
+     * All available marshallers are consulted in sequence, in ascending numerical priority. The first (lowest numerical
+     * priority) marshaller to accept an object is used.
+     * </p>
+     * 
+     * @return the priority of this marshaller
+     */
+    int priority();
+
+    /**
+     * Does this marshaller accept the given export?
+     * 
+     * @param export the object to test
+     * @return true if this marshaller should be used, false to use the remainder of the marshaller chain
+     */
+    boolean accept(Object export);
+
+    /**
+     * Processes a snapshot request using the provided parameters.
+     *
+     * @param snapshotRequest the request specifying details for the snapshot operation
+     * @param options the configuration options for the snapshot, such as batch size or message size limits
+     * @param export the object to be exported as part of the snapshot
+     * @param metrics helper for recording and reporting snapshot performance metrics
+     * @param listener the observer to handle messages generated during the snapshot process
+     * @param ticketLogName the name of the ticket log associated with the snapshot request
+     * @param streamGeneratorFactory the factory to generate message streams for the snapshot operation
+     */
+    void snapshot(BarrageSnapshotRequest snapshotRequest,
+            BarrageSnapshotOptions options,
+            Object export,
+            BarragePerformanceLog.SnapshotMetricsHelper metrics,
+            StreamObserver<BarrageMessageWriter.MessageView> listener, String ticketLogName,
+            BarrageMessageWriter.Factory streamGeneratorFactory);
+
+    /**
+     * Subscribes to a Barrage stream using the provided subscription request and options.
+     *
+     * @param subscriptionRequest the request specifying the details of the subscription
+     * @param options the options to configure the subscription
+     * @param export the exported object to be subscribed
+     * @param listener the listener to handle messages generated by the subscription
+     * @return a Subscription object to manage the lifecycle of the Barrage subscription
+     */
+    Subscription subscribe(BarrageSubscriptionRequest subscriptionRequest,
+            BarrageSubscriptionOptions options,
+            Object export,
+            StreamObserver<BarrageMessageWriter.MessageView> listener);
+
+    /**
+     * The return value from a
+     * {@link #subscribe(BarrageSubscriptionRequest, BarrageSubscriptionOptions, Object, StreamObserver)} call.
+     */
+    interface Subscription {
```

Comment:
Is there value in making the Subscription a LivenessReferent explicitly (instead of `#toManage`)? Could it enforce better/more-straightforward session-expiration cleanup, generically speaking?

Reply:
ExportObject's can hold references to static objects (like static tables), but are always considered LivenessReferents as they have a well-defined life cycle. I'm thinking of Subscription instances in this context -- created for a specific gRPC call and has a tied lifetime to the gRPC call. 

That said, I guess the outer Marshaller is a SingletonLivenessManager that the SessionState is to destroy via:
```java
            this.session.addOnCloseCallback(this);
```

So, nevermind, this is certainly a reasonable ownership model.